cookbook/controller/upload_file.rst

```diff
@@ -0,0 +1,161 @@
+.. index::
+   single: Controller; Upload; File
+
+How to Upload Files
+===================
+
+.. note::
+
+    Instead of handling file uploading yourself, you may consider using the
+    `VichUploaderBundle`_ community bundle. This bundle provides all the common
+    operations (such as file renaming, saving and deleting) and it's tightly
+    integrated with Doctrine ORM, MongoDB ODM, PHPCR ODM and Propel.
+
+Imagine that you have a ``Product`` entity in your application and you want to
+add a PDF brochure for each product. To do so, add a new property called ``brochure``
+in the ``Product`` entity::
+
+    // src/AppBundle/Entity/Product.php
+    namespace AppBundle\Entity;
+
+    use Doctrine\ORM\Mapping as ORM;
+    use Symfony\Component\Validator\Constraints as Assert;
+
+    class Product
+    {
+        // ...
+
+        /**
+         * @ORM\Column(type="string")
+         *
+         * @Assert\NotBlank(message="Please, upload the product brochure as a PDF file.")
+         * @Assert\File(mimeTypes={ "application/pdf" })
+         */
+        private $brochure;
+
+        public function getBrochure()
+        {
+            return $this->brochure;
+        }
+
+        public function setBrochure($brochure)
+        {
+            $this->brochure = $brochure;
+
+            return $this;
+        }
+    }
+
+Note that the type of the ``brochure`` column is ``string`` instead of ``binary``
+or ``blob`` because it just stores the PDF file name instead of the file contents.
+
+Then, add a new ``brochure`` field to the form that manages ``Product`` entities::
+
+    // src/AppBundle/Form/ProductType.php
+    namespace AppBundle\Form;
+
+    use Symfony\Component\Form\AbstractType;
+    use Symfony\Component\Form\FormBuilderInterface;
+
+    class ProductType extends AbstractType
+    {
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                // ...
+                ->add('brochure', 'file', array('label' => 'Brochure (PDF file)'))
+                // ...
+            ;
+        }
+
+        // ...
```

Comment:
Let's include the `setDefaultOptions()` here with `data_class`, and also `getName()`. I think it'll just make the entry "complete"